src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+ *
+ * Server monitor must be in state MONGOC_THREAD_OFF.
+ */
+static BSON_THREAD_FUN (_server_monitor_thread, server_monitor_void)
+{
+   mongoc_server_monitor_t *server_monitor;
+   mongoc_server_description_t *description;
+   mongoc_server_description_t *previous_description;
+
+   server_monitor = (mongoc_server_monitor_t *) server_monitor_void;
+   description =
+      mongoc_server_description_new_copy (server_monitor->description);
+   previous_description = NULL;
+
+   while (true) {
+      bool cancelled = false;
+
+      bson_mutex_lock (&server_monitor->shared.mutex);
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         bson_mutex_unlock (&server_monitor->shared.mutex);
+         break;
+      }
+      bson_mutex_unlock (&server_monitor->shared.mutex);
+
+      mongoc_server_description_destroy (previous_description);
+      previous_description = mongoc_server_description_new_copy (description);
+      mongoc_server_description_destroy (description);
+      description = mongoc_server_monitor_check_server (
+         server_monitor, previous_description, &cancelled);
+
+      if (cancelled) {
+         mongoc_server_monitor_wait (server_monitor);
+         continue;
+      }
+
+      _server_monitor_update_topology_description (server_monitor, description);
+
+      /* Immediately proceed to the next check if the previous response was
+       * successful and included the topologyVersion field. */
+      if (description->type != MONGOC_SERVER_UNKNOWN &&
+          !bson_empty (&description->topology_version)) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to topologyVersion");
+         continue;
+      }
+
+      /* ... or the previous response included the moreToCome flag */
+      if (server_monitor->more_to_come) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to moreToCome");
+         continue;
+      }
+
+      /* ... or the server has just transitioned to Unknown due to a network
+       * error. */
+      if (_mongoc_error_is_network (&description->error) &&
+          previous_description->type != MONGOC_SERVER_UNKNOWN) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to network error");
+         continue;
+      }
+
+      mongoc_server_monitor_wait (server_monitor);
+   }
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.state = MONGOC_THREAD_JOINABLE;
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+   mongoc_server_description_destroy (previous_description);
+   mongoc_server_description_destroy (description);
+   BSON_THREAD_RETURN;
+}
+
+static bool
+_server_monitor_ping_server (mongoc_server_monitor_t *server_monitor,
+                             int64_t *rtt_ms)
+{
+   bool ret = false;
+   int64_t start_us = _now_us ();
+   bson_t ismaster_reply;
+   bson_error_t error;
+
+   *rtt_ms = -1;
```

Comment:
Is `-1` a legitimate return value if we cannot conduct a ping? Just want to make sure it's handled properly elsewhere. Assuming this relates to `_server_monitor_rtt_thread` below, I might suggest using a defined constant like `RTT_UNSET` instead of `-1` for self-documenting purposes.